### ISSUE TITLE
Fix S2583 FP: if condition after a for loop within a loop

### ISSUE THREAD

**Issue by llfab** [SonarSource/sonar-dotnet#8833]
### Description

False positive detected by scanner: csharpsquid:S2583

### Repro steps

![image](https://github.com/SonarSource/sonar-dotnet/assets/31031996/2976f66a-9feb-4d74-bdf1-57a36205a860)

or 

![image](https://github.com/SonarSource/sonar-dotnet/assets/31031996/0594f5c6-b8b1-4d24-acc5-640bca048ae0)

### Expected behavior

This should not be a detection at all

### Actual behavior

Currently, this is detected as csharpsquid:S2583

### Related information

* C# 12
* Visual Studio 2022
* .net 8.0
* SonarScanner for MSBuild 5.13.1
* Windows

**Comment by sebastien-marichal** [comment]
Hello @llfab,

Thank you for reporting this issue.

Can you provide a text-based reproducer?
It is hard for us to investigate with images.

Also, which version of the C# plugin are you using?
You should be able to get this information from the SonarScanner for MSBuild log. It should look like this:
```text
12:03:00.04  Processing plugin: csharp version 9.20.0.85982
```

We recently made important changes to our Symbolic Execution engine targeting this kind of false positive within loops. Those changes are available in [9.20](https://github.com/SonarSource/sonar-dotnet/releases/tag/9.20.0.85982).

**Comment by llfab** [comment]
This is one sample:

```
        public MeshBuilder3D AddArrow(Point3D anchor, Point3D tip, double stemRadius,
            double tipRadius, double tipHeight, Pose3D additionalPose)
        {
            double height = Point3D.Distance(anchor, tip);
            double[] R = { stemRadius, stemRadius, tipRadius,
                0.00001 * tipRadius, // auxiliary loop (instead of pointy tip) prevents rendering artifacts
            };
            double[] Z = { -height, -tipHeight, -tipHeight, 0 };

            ShapeIndexer indexer = new ShapeIndexer();

            Point3D tipCenter = new Point3D(0, 0, 0);
            Point3D baseCenter = new Point3D(0, 0, -height);

            Point3D[] q = null;
            foreach (double phi in new Interval(0, 2 * Math.PI).Discretize(17))
            {
                Point3D[] p = new Point3D[4];
                for (int j = 0; j < p.Length; j++) { p[j] = new Point3D(R[j] * Math.Cos(phi), R[j] * Math.Sin(phi), Z[j]); }

                if (q != null)
                {
                    indexer.AddTriangle(p[0], q[0], baseCenter);
                    for (int j = 0; j < 3; j++) { indexer.AddQuad(q[j], p[j], p[j + 1], q[j + 1]); }
                    indexer.AddTriangle(q[3], p[3], tipCenter);
                }

                q = p;
            }

            Pose3D pose = Pose3D.CreatePoseFromTo(Point3D.Origin, Vector3D.UnitZ, Vector3D.UnitY,
                tip, tip - anchor, (tip - anchor).CreateOrthogonalVector());

            Add(indexer.GetShape(), pose);

            return this;
        }
```

**Comment by llfab** [comment]
This is the other sample

        ```
public static Pose3D ReadTfm(Stream stream)
        {
            StreamReader reader = new StreamReader(stream, true);
            string[] lines = reader.ReadToEnd().
                Split(new[] { Environment.NewLine }, StringSplitOptions.None).
                Select(l => l.Trim()).
                Where(l => !string.IsNullOrWhiteSpace(l) && !l.StartsWith("#")).ToArray();

            Assert.That(lines.Length == 5, "TFM parser error!");
            Assert.That(lines[4] == "Units: mm", "TFM parser error!");

            List<double> M = new List<double>();
            foreach (string line in lines.Take(4))
            {
                string[] parts = line.Split(new[] { " " }, StringSplitOptions.RemoveEmptyEntries);
                Assert.That(parts.Length == 4, "TFM parser error!");
                for (int i = 0; i < 4; i++)
                {
                    double val;
                    bool ok = double.TryParse(parts[i], CultureInfo.InvariantCulture, out val);
                    Assert.That(ok, "TFM parser error!");
                    M.Add(val);
                }
            }

            #region Validate Matrix
            Assert.That(M[12] == 0 && M[13] == 0 && M[14] == 0 && M[15] == 1,
                "TFM parser error - transformation is non-rigid!");

            double[] rotation = new[] { M[0], M[1], M[2], M[4], M[5], M[6], M[8], M[9], M[10] };
            for (int i = 0; i < 3; i++)
            {
                for (int j = i; j < 3; j++)
                {
                    double dot = 0;
                    for (int k = 0; k < 3; k++) { dot += rotation[3 * k + i] * rotation[3 * k + j]; }
                    Assert.That(Accuracy.AreEqual(dot, (i == j) ? 1 : 0, Accuracy.EpsilonLow),
                        "TFM parser error - transformation is non-rigid");
                }
            }
            #endregion

            return new Pose3D(rotation, M[3], M[7], M[11]);
        }
```

**Comment by llfab** [comment]
Scanner versions:

```
2024-03-07T02:36:13.2974757+01:00 |   SonarScanner for MSBuild 6.2
2024-03-07T02:36:13.3048013+01:00 |   Using the .NET Framework version of the Scanner for MSBuild
2024-03-07T02:36:13.5007202+01:00 |   Pre-processing started.
2024-03-07T02:36:13.5032025+01:00 |   Preparing working directories...
2024-03-07T02:36:13.5165044+01:00 |   02:36:13.516  Updating build integration targets...
2024-03-07T02:36:13.6572028+01:00 |   02:36:13.657  Fetching analysis configuration settings...
2024-03-07T02:36:14.0676103+01:00 |   02:36:14.067  Provisioning analyzer assemblies for cs...
2024-03-07T02:36:14.0794759+01:00 |   02:36:14.067  Installing required Roslyn analyzers...
2024-03-07T02:36:14.0801436+01:00 |   02:36:14.067  Processing plugin: csharp version 9.21.0.86780
2024-03-07T02:36:14.2392633+01:00 |   02:36:14.239  Processing plugin: vbnet version 9.21.0.86780
2024-03-07T02:36:14.3[645](https://gitlab.com/strykercorp/t-and-e/EnTech/BitsOfNature/-/jobs/6336946850#L645)426+01:00 |   02:36:14.364  Processing plugin: securitycsharpfrontend version 10.5.0-M2.30409
2024-03-07T02:36:14.7086270+01:00 |   02:36:14.708  Provisioning analyzer assemblies for vbnet...
2024-03-07T02:36:14.71[648](https://gitlab.com/strykercorp/t-and-e/EnTech/BitsOfNature/-/jobs/6336946850#L648)31+01:00 |   02:36:14.708  Installing required Roslyn analyzers...
2024-03-07T02:36:14.7169994+01:00 |   02:36:14.708  Processing plugin: csharp version 9.21.0.86780
2024-03-07T02:36:14.7175099+01:00 |   02:36:14.708  Processing plugin: vbnet version 9.21.0.86780
2024-03-07T02:36:14.7195543+01:00 |   02:36:14.719  Incremental PR analysis: Base branch parameter was not provided.
2024-03-07T02:36:14.7195543+01:00 |   02:36:14.719  Cache data is empty. A full analysis will be performed.
2024-03-07T02:36:14.7313089+01:00 |   02:36:14.731  Pre-processing succeeded.
2024-03-07T02:36:14.7404486+01:00 |   COMPLETED: Start Scanner (MASTER BRANCH)
```

**Comment by sebastien-marichal** [comment]
Hello @llfab,

I was able to reproduce the issue.
It seems the engine is having difficulties when the condition is preceded by a for loop. Maybe with other loops as well, I didn't test it.

```cs
var i = 0;
while (i < 3)
{
    var j = 0;
    while (j < 3)
    {
        // Removing this loop will make S2583 disappear
        for (var k = 0; k < 3; k++)
        {
            Console.WriteLine("Loop!");
        }

        Console.WriteLine((i == j) ? 1 : 0); // S2583 FP on (i == j)
        j++;
    }

    i++;
}
```

**Comment by sebastien-marichal** [comment]
Turns out the issue is already fixed on the master branch.

I was able to reproduce this issue on 9.21.0.86780.

@llfab the fix will be available in 9.22.0.
